### ISSUE TITLE
Non-raw string checking in python regex interface

### ISSUE THREAD

**Issue by mpacer** [firasdib/Regex101#880]
Short version: It would be nice to have the option to turn off the expectation that the input string will be a raw string. 

In particular this arises in the context of you working with regular expressions that have been created by someone else in a programmatic fashion (i.e., are never encoded as raw literal strings).

This service is awesome for exploring the meaning & interpretation of regular expressions, but in cases where you are debugging an extant regular expression that already exists in memory `my_exp.pattern` will successfully give you *a* string, but there is no straightforward way to transform that string into what would be its raw literal equivalent.  

Fixing these weird escaping edge cases can be really slow and tiresome. Rather than requiring this effort, it'd be much nicer if there were a way to simply create a string literal instead of a raw string literal!

Long version (with example):

I have been trying to debug some issues that have arisen around https://github.com/lepture/mistune & https://github.com/jupyter/nbconvert. 

Mistune programmatically creates many regular expressions 

After a while, I figured out how to get the pattern that was causing an issue. As a string obtained via `rule.pattern` it was:

```python
>>> rule.pattern
'^ *(?:<!--[\\s\\S]*?-->|<((?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img|font)\\b)\\w+?(?!:/|[^\\w\\s@]*@)\\b)((?:\\s*[a-zA-Z\\-](?:\\=(?:"[^"]*"|\'[^\']*\'|[^\\s\'">]+))?)*?)>([\\s\\S]*?)<\\/\\1>|<(?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img|font)\\b)\\w+?(?!:/|[^\\w\\s@]*@)\\b(?:\\s*[a-zA-Z\\-](?:\\=(?:"[^"]*"|\'[^\']*\'|[^\\s\'">]+))?)*?\\s*\\/?>) *(?:\\n{2,}|\\s*$)'
```

The problem is that if you were to express this in terms of it's raw literal value, which copying and pasting that into the current regex101.com interface, this would be interpreted as 

```python
>>> r'^ *(?:<!--[\\s\\S]*?-->|<((?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img|font)\\b)\\w+?(?!:/|[^\\w\\s@]*@)\\b)((?:\\s*[a-zA-Z\\-](?:\\=(?:"[^"]*"|\'[^\']*\'|[^\\s\'">]+))?)*?)>([\\s\\S]*?)<\\/\\1>|<(?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img|font)\\b)\\w+?(?!:/|[^\\w\\s@]*@)\\b(?:\\s*[a-zA-Z\\-](?:\\=(?:"[^"]*"|\'[^\']*\'|[^\\s\'">]+))?)*?\\s*\\/?>) *(?:\\n{2,}|\\s*$)'
'^ *(?:<!--[\\\\s\\\\S]*?-->|<((?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img|font)\\\\b)\\\\w+?(?!:/|[^\\\\w\\\\s@]*@)\\\\b)((?:\\\\s*[a-zA-Z\\\\-](?:\\\\=(?:"[^"]*"|\\\'[^\\\']*\\\'|[^\\\\s\\\'">]+))?)*?)>([\\\\s\\\\S]*?)<\\\\/\\\\1>|<(?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img|font)\\\\b)\\\\w+?(?!:/|[^\\\\w\\\\s@]*@)\\\\b(?:\\\\s*[a-zA-Z\\\\-](?:\\\\=(?:"[^"]*"|\\\'[^\\\']*\\\'|[^\\\\s\\\'">]+))?)*?\\\\s*\\\\/?>) *(?:\\\\n{2,}|\\\\s*$)'
```

Which is wrong, it's actually treating the `\\` as unescaped `\`'s therefore expecting all cases that should be `\` (for example in defining shorthand character classes like `\s`) to be `\\`.

One way to get at this would be to attempt to print it, but then that introduced the issue of unescaping punctuation marks (e.g.,  `\'` becoming `'`) which meant that regex was no longer a valid python string if delimited by either `'` or `"`:

```python
>>> print(rule.pattern)
^ *(?:<!--[\s\S]*?-->|<((?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img|font)\b)\w+?(?!:/|[^\w\s@]*@)\b)((?:\s*[a-zA-Z\-](?:\=(?:"[^"]*"|'[^']*'|[^\s'">]+))?)*?)>([\s\S]*?)<\/\1>|<(?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img|font)\b)\w+?(?!:/|[^\w\s@]*@)\b(?:\s*[a-zA-Z\-](?:\=(?:"[^"]*"|'[^']*'|[^\s'">]+))?)*?\s*\/?>) *(?:\n{2,}|\s*$)
```

To fix this required manually editing the printed output to reëscape the punctuation (specifically `'` and `"`) or noticing that there were no doubled single-quotes (`''…''`) or doubled double-quotes (`""…""`) and changing the delimiter to match that. 

Rather than requiring this effort, it'd be much nicer if there were a way to simply create a string literal instead of a raw string literal, since I could have then just directly copied and pasted the existing string literal representation. It took a lot of time looking up how to "recreate" what would have been the raw string literal content and I'd like no one else to have to do that if it can be avoided.

**Comment by TWiStErRob** [comment]
Use print to get the regex out of your language, no escaping should be included, because that's not part of the regex language, it is not aware of its representation and shouldn't be. Double-quotes, double-edged are part of Python, not the regex flavor Python supports.

To debug a regex:
* print it and copy
* fix the problems
* use Code Generator to re-escape it

Another thing to note is that the language has a raw string literal construct, if you are using that, then there's no escaping issue as I see. It's copy between `r'...'` and then paste back.

Also see: https://stackoverflow.com/a/1732454/253468 ;)

**Comment by mpacer** [comment]
All that you said I understand. I've debugged many regexes — it's why I know, love and am trying to contribute to the development of this service. I'm pointing out weaknesses with the approach of exclusively allowing raw literal strings to be debugged in the web interface.

Most notably, while the language has a literal syntax construct it is merely a convenience syntax and does not represent a different kind of underlying object. This means that printing will not always work to recreate what would have been the raw string, especially once delimiters are in the picture. 

Printing and copying won't work if, for example the regex has an escaped version of each kind of quote in a non-raw string because it was constructed out of a concatenation (or other manipulation) of other raw strings. For example,

```
a = r'"""|(?:"?)'
b = r"'''|(?:'?)"
c = a+"|"+b
```
There would be no way to express c using a single raw string without escaping delimiters given the way that python delimiting works. However you could using a standard literal string, for example 

```
d = "\"\"\"|(?:\")|'''|(?:'?)"
```

This is not part of regex escaping, it has only to do with python string notation. 

One could not test the examples I just gave using raw strings because all of the `\` would be interpreted as literal '\' characters(`'\\'`) rather than as the python escape characters that they are intended to be (except see below regarding the delimiter of the string itself). However those should be valid strings in python however that internal representation is created (either using separate raw literals or fully escaped strings).

The site already allows escapement of this variety in a raw string in order to be able to include the delimiter in question in the string. It just means the semantics of the string depend on the delimiters you choose. E.g., 

```
e = '\'\'\'|(?:\'?)'
```

All I'm asking is that you allow the possibility of testing with a non-raw literal string in the web interface. That allows hopping over all of these delimiter specific escapement issues. It also makes it easier when working in a py2/py3 compatible codebase where the notion of a string that is raw and explicitly made to be Unicode at the same time (denoted `ur''`) can't be used. 

I still think the default should still be to include the raw option. But for some cases it's _much_ more straightforward to be able to use standard python strings including standard python string escapement. It comes up often enough  especially in really complicated cases (those where regex101's introspection abilities are most powerful) that it would be well worth it.

**Comment by TWiStErRob** [comment]
@firasdib it sounds like the site needs a code parser to be able to read back all code generated by the site.
I guess an unescape button would be a good solution the question is how to know what type of delimiter was used.

**Comment by Doqnach** [comment]
I think this feature would be in line with other "quick fix" requests like #865 and #870

**Comment by firasdib** [comment]
Im not convinced this is something that should be supported, because its supposed to be a regex tester (not tailored to any specific language). I also don't see a trivial way of doing what you're asking, this would require me to rework the regex parser and stuff.

For now, this won't be added.
